pkg/reconciler/apis/permissionclaimlabel/permissionclaimlabel_reconcile.go

```diff
@@ -0,0 +1,248 @@
+		return nil
+	}
+
+	exportClusterName := apiBinding.Status.BoundAPIExport.Workspace.Path
+	exportName := apiBinding.Status.BoundAPIExport.Workspace.ExportName
+	apiExport, err := c.getAPIExport(logicalcluster.New(exportClusterName), exportName)
+	if err != nil {
+		logger.Error(err, "error getting APIExport", "apiExportWorkspace", exportClusterName, "apiExportName", exportName)
+		return nil // nothing we can do
+	}
+
+	logger = logger.WithValues(logging.FromPrefix("apiExport", apiExport)...)
+
+	exportedClaims := sets.NewString()
+	for _, claim := range apiExport.Spec.PermissionClaims {
+		exportedClaims.Insert(setKeyForClaim(claim))
+	}
+
+	acceptedClaims := sets.NewString()
+	for _, claim := range apiBinding.Spec.PermissionClaims {
+		if claim.State == apisv1alpha1.ClaimAccepted {
+			acceptedClaims.Insert(setKeyForClaim(claim.PermissionClaim))
+		}
+	}
+
+	appliedClaims := sets.NewString()
+	for _, claim := range apiBinding.Status.AppliedPermissionClaims {
+		appliedClaims.Insert(setKeyForClaim(claim))
+	}
+
+	expectedClaims := exportedClaims.Intersection(acceptedClaims)
+	unexpectedClaims := acceptedClaims.Difference(expectedClaims)
+	needToApply := expectedClaims.Difference(appliedClaims)
+	needToRemove := appliedClaims.Difference(acceptedClaims)
+	allChanges := needToApply.Union(needToRemove)
+
+	logger.V(6).Info("claim set details",
+		"expected", expectedClaims,
+		"unexpected", unexpectedClaims,
+		"toApply", needToApply,
+		"toRemove", needToRemove,
+		"all", allChanges,
+	)
+
+	var allErrs []error
+	applyErrors := sets.NewString()
+
+	for _, s := range allChanges.List() {
+		claim := claimFromSetKey(s)
+		claimLogger := logger.WithValues("claim", s)
+
+		informer, gvr, err := c.getInformerForGroupResource(claim.Group, claim.Resource)
+		if err != nil {
+			allErrs = append(allErrs, fmt.Errorf("error getting informer for group=%q, resource=%q: %w", claim.Group, claim.Resource, err))
+			if acceptedClaims.Has(s) {
+				applyErrors.Insert(s)
+			}
+			continue
+		}
+
+		claimLogger.V(4).Info("listing resources")
+		objs, err := informer.Informer().GetIndexer().ByIndex(indexers.ByLogicalCluster, clusterName.String())
+		if err != nil {
+			allErrs = append(allErrs, fmt.Errorf("error listing group=%q, resource=%q: %w", claim.Group, claim.Resource, err))
+			if acceptedClaims.Has(s) {
+				applyErrors.Insert(s)
+			}
+			continue
+		}
+
+		claimLogger.V(4).Info("got resources", "count", len(objs))
+
+		var claimErrs []error
+		for _, obj := range objs {
+			objectMeta, err := meta.Accessor(obj)
+			if err != nil {
+				claimErrs = append(claimErrs, fmt.Errorf("error getting meta accessor for %T: %w", obj, err))
+				continue
+			}
+
+			patchLogger := claimLogger.WithValues("objectNamespace", objectMeta.GetNamespace(), "objectName", objectMeta.GetName())
+			patchLogger.V(4).Info("patching to get claim labels updated")
+
+			// Empty patch, allowing the admission plugin to update the resource to the correct labels
```

Comment:
So is it client-side that stops no-op patches?

Reply:
In this case, because the controller has determined an update is needed, when the empty patch is issued, the mutating admission controller mutates the incoming object first